PLAT-72723: Update ui/VirtualList lifecycle method

ybsung opened the pull request:
### Checklist

* [X] I have read and understand the [contribution guide](http://enactjs.com/docs/developer-guide/contributing/)
* [ ] A [CHANGELOG entry](http://enactjs.com/docs/developer-guide/contributing/changelogs/) is included
* [ ] At least one test case is included for this feature or bug fix
* [X] Documentation was added or is not needed
* [ ] This is an API breaking change

### Issue Resolved / Feature Added
[//]: # (Describe the issue resolved or feature added by this pull request)

`UNSAFE_componentWillMount()`, `UNSAFE_componentWillReceiveProps()`, and `UNSAFE_componentWillUpdate()` will be deprecated.

### Resolution
[//]: # (Does the code work as intended?)
[//]: # (What is the impact of this change and *why* was it made?)

- Move the code in `UNSAFE_componentWillMount()` to `constructor`.
But `updateStatesAndBounds` was renamed as `getStatesAndUpdateBounds` to get a new state instead of calling `setState` in the `updateStatesAndBounds`.

- Move the code in `UNSAFE_componentWillReceiveProps()` to `componentDidUpdate`.
But we still need the `this.hasDataSizeChanged` variable used in `moonstone/VirtualList`.
In addition, `this.cc = []` for RTL was removed.

- Move the code in `UNSAFE_componentWillUpdate()` and the part of the `positionItems` in render phase to `getDerivedStateFromProps` so that the previous first index and the previous `childProps` are only referenced in the `getDerivedStateFromProps`.

### Links
[//]: # (Related issues, references)

PLAT-72723

ybsung commented on packages/ui/VirtualList/VirtualListBase.js:
```diff
@@ -666,6 +637,11 @@ const VirtualListBaseFactory = (type) => {
+			this.prevFirstIndex = (
+				this.prevFirstIndex === this.state.firstIndex ||
+				(this.prevProps.childProps || this.props.childProps) && this.prevProps.childProps !== this.props.childProps
+			) ? -1 : this.prevFirstIndex;
```
The code above is moved from the following code in `UNSAFE_componentWillUpdate()`.
```
UNSAFE_componentWillUpdate (nextProps, nextState) {
	if (this.state.firstIndex === nextState.firstIndex || this.props.childProps && this.props.childProps !== nextProps.childProps) {
		this.prevFirstIndex = -1; // force to re-render items
	}
}
```

ybsung commented on packages/ui/VirtualList/VirtualListBase.js:
```diff
@@ -684,8 +660,8 @@ const VirtualListBaseFactory = (type) => {
-				{primaryPosition, secondaryPosition} = this.getGridPosition(updateFrom),
-				width, height;
+				width, height,
+				{primaryPosition, secondaryPosition} = this.getGridPosition(updateFrom);
```
I fixed it due to `unexpected reserved word let`.

ybsung commented on packages/ui/VirtualList/VirtualListBase.js:
```diff
@@ -797,10 +771,42 @@ const VirtualListBaseFactory = (type) => {
 				this.positionItems();
 			}
 
+			const {cc} = this;
```
I could not replace `cc` with`this.cc` in `{itemsRenderer({cc, initItemContainerRef, primary})}` due to `unexpected reserved word this`.

ybsung commented on packages/ui/VirtualList/VirtualListBase.js:
```diff
@@ -797,10 +771,42 @@ const VirtualListBaseFactory = (type) => {
+					this.setScrollPosition(x, y, 0, 0, props.rtl);
+				}
+			}
+
```
I just moved the code in `UNSAFE_componentWillReceiveProps()` to here.

0x64 commented:
LGTM

viodragon2 commented on packages/ui/VirtualList/VirtualListBase.js:
```diff
@@ -797,14 +771,44 @@ const VirtualListBaseFactory = (type) => {
+
+			this.hasDataSizeChanged = (dataSize !== props.dataSize);
+
+			if (hasMetricsChanged || this.props.clientSize && this.primary === null && this.secondary === null) {
```
All these conditions seem to happen after the first render. IOW, metrics and data size change would only happen after its first mount and subsequent renders.

I wonder if there are any noticeable perf loss if these changes were pushed back to `componentDidUpdate` phase. By doing so, I think we can remove the hacky `updatePrevProps()` and (yet another) the internal member variable. Plus, we can remove the direct `state` mutation hack that I pointed out above.

ybsung replied:
The following condition is for the first render.
- this.props.clientSize && this.primary === null && this.secondary === null

The original code  for the condition above was in `UNSAFE_componentWillMount ()`. The reason why we added the condition was to reduce calling `render()` from twice to once to enhance app launch time. The impact for the patch was so big. Please check the comments in PLAT-37236. If it is ok to call `render()` more, we could just undo the patch by removing the condition. But I'm afraid of increasing the launch time.

viodragon2 replied:
Generic questions:

- Have we promoted `clientSize` props over setting dimensions to developers? I don't see any apps using this, unfortunately :(
- I don't think we can't ignore what happens when VirtualList is used with `Scroller` as `VirtualListBase` has a strong relationship with `Scrollable`. And `syncClientSize` gets called when `Scrollable` updates.
- If the goal is to minimize the initial render when `clientSize` were provided as a prop, can we calculate metrics in `constructor` instead? Essentially we're initializing the metrics via `calculateMetrics` so that we can position items on render.

ybsung replied:
> * Have we promoted `clientSize` props over setting dimensions to developers? I don't see any apps using this, unfortunately :(

I found just only one use case in Settings app. Actually the issue related with `clientSize` was raised from Settings Picture panel. But the `clientSize` in the panel was not used.
\com.webos.app.settings\src\views\General\LanguageList.js

```
	setVirtualListProps (listData, comp) {
		const retProps = {
			itemSize: ri.scale(78),
			dataSize: listData.length,
			isItemDisabled: (index) => (listData[index].disabled),
			itemRenderer: comp
		};

		if (typeof window === 'object' && window.testing) {
			retProps.clientSize = {
				clientWidth: 1920,
				clientHeight: 1080
			};
		}

		return retProps;
	}

	const virtualListProps = this.setVirtualListProps(listData, this.getLanguageListItem());

	<VirtualList {...virtualListProps} className={css['general-language-list-virtuallist']} />
```

ybsung replied:
> * If the goal is to minimize the initial render when `clientSize` were provided as a prop, can we calculate metrics in `constructor` instead? Essentially we're initializing the metrics via `calculateMetrics` so that we can position items on render.

Yes. I think it makes sense. As you proposed the code in [your branch](https://github.com/enactjs/enact/tree/feature/PLAT-72723-stephen) I'll try to apply it.

ybsung replied:
<On Feb 18th>
- I move the code in `UNSAFE_componentWillMount()` to `constructor` related with `clientSize`.
But `updateStatesAndBounds` was renamed as `getStatesAndUpdateBounds` to get a new state because we could not call `setState` in `constructor`.

ybsung commented on packages/ui/VirtualList/VirtualListBase.js:
```diff
@@ -797,14 +771,44 @@ const VirtualListBaseFactory = (type) => {
+
+				this.cc = [];
+				if (type === Native) {
+					this.scrollToPosition(x, y, props.rtl);
```
If `rtl` mode changes, we need to update `VirtualList` as well as `VirtualList items` based on the `rtl` mode.

For `VirtualList`, we called the `this.scrollToPosition` or `this.setScrollPosition` here. There might be no problem if calling it in `componentDidUpdate`.

But for `VirtualList items`, `this.cc = [];` was executed at https://github.com/enactjs/enact/blob/31e372d4d38972d7e0448a0f4481a6502dde2efe/packages/ui/VirtualList/VirtualListBase.js#L796, then `this.positionItems();` was called at https://github.com/enactjs/enact/blob/31e372d4d38972d7e0448a0f4481a6502dde2efe/packages/ui/VirtualList/VirtualListBase.js#L805 If moving the `this.cc =[];` to `componentDidUpdate`,
- To render the items, we need to call `this.forceUpdate()` instead of `this.setState()` in `componentDidUpdate` because no state changes. I think calling `forceUpdate` is not a good pattern.
- `render()` function will be called once more.

So I thought it would be better to execute `this.cc = [];` in `render`() with calling `scrollTo` together.

viodragon2 replied:
Is `this.cc` required to be reset when `rtl` prop changes? Doesn't index and key stay the same in RTL layout? Or am I missing something?

ybsung replied:
All items usually display properly depending on the RTL mode based on CSS. But app developers could position the content in the item depending on the RTL mode. For this case, we need to re-render the items so that app developers could position them properly in `itemRender`.

ybsung replied:
But there is no real case until now. Let's move this logic to `componentDidUpdate`.

ybsung replied:
<On Feb 18th>
- The code related with RTL including calling `scrollTo` is moved to not render phase but `componentDidUpdate`.
- The code `this.cc = [];` that we discussed is removed. If any issue raised, then we could revisit it.

ybsung commented on packages/ui/VirtualList/VirtualListBase.js:
```diff
@@ -448,7 +414,12 @@ const VirtualListBaseFactory = (type) => {
 
-			this.setState({firstIndex: newFirstIndex, numOfItems});
+			if (slient) {
+				this.state.firstIndex = newFirstIndex; // eslint-disable-line react/no-direct-mutation-state
```
I'll try to find a better way.

ybsung replied:
<On Feb 18th>
The code to change the state directly in `calculateMetrics` was removed.
`getStatesAndUpdateBounds` ( which name was `updateStateAndBounds` before ) returns a new state instead of updating the state directly.

ybsung commented:
Hi @viodragon2 
There are lots of bad practices in `VirtualList`. But to enhance the launch time performance and the scroll performance, we manipulate the DOM elements and the states manually in the `VirtualList`. As I mentioned the comments in PLAT-37236, the impact of calling the render function once more in `VirtualList` seems to be pretty large.

IMHO, if possible, we need to reduce calling the render function in `VirtualList`. To do that, we could not avoid updating states directly and referring the previous props from states in render phase.

But it is true that it is the right time to remove all the bad practices. Then we can do improve performance from another perspective for several months.

Please let me know your thought.

viodragon2 commented on packages/ui/VirtualList/VirtualListBase.js:
```diff
@@ -797,14 +771,44 @@ const VirtualListBaseFactory = (type) => {
+
+			if (hasMetricsChanged || this.props.clientSize && this.primary === null && this.secondary === null) {
+				this.calculateMetrics(props);
+				this.updateStatesAndBounds(props, true);
```
As the name of the function suggests, I wonder if we can translate this to `getDerivedStatesFromProps` so that we don't worry about directly mutating states.

This would require for `cbScrollTo` it to be called elsewhere (maybe in `componentDidUpdate`?) and I'm not sure if that would cause a too big of a change.

What do you think?

ybsung replied:
>As the name of the function suggests, I wonder if we can translate this to `getDerivedStatesFromProps` so that we don't worry about directly mutating states.

As I mentioned in the JIRA, my first approach is the one that you mentioned. I moved all the code in `componentWillReceiveProps` to `getDrivedStateFromProps`.

As you can see the following code, I renamed `updateStateAndBounds` as `updateMetrics`, passed `prevProps`, `props`, and `prevState` to the function, then the function returned a new state. To do that, I needed to make all functions called from `updateStateAndBounds` returning a new state.
https://github.com/enactjs/enact/compare/feature/PLAT-72723-enhanced2#diff-decb3acaa813ab261efba3ee57c8b4f9R235

But I thought we needed to update lots of code to change member variables to state variables and member functions to class static functions or global functions because `getDrivedStateFromProps` was a static function and only props and states were passed.

If it is the right way, I'll keep going ahead instead of this PR.

ybsung replied:
<On Feb 18th>
- The code above including `cbScrollTo` is moved to `componentDidUpdate`.

viodragon2 commented on packages/ui/VirtualList/VirtualListBase.js:
```diff
@@ -797,14 +771,44 @@ const VirtualListBaseFactory = (type) => {
+				this.setContainerSize();
+			} else if (this.hasDataSizeChanged) {
+				this.updateStatesAndBounds(props, true);
+				this.setContainerSize();
```
Why do we need to reset container size when only the `dataSize` has changed?

ybsung replied:
I just copied my comment in https://github.com/enactjs/enact/pull/2108#discussion_r257104333.
```
In addition, when metrics changes or data size changes, we need to update container size again because the visibility of the scrollbar could changes and the scroll area could changes.
```

ybsung replied:
<On Feb 18th>
- I updated the PR.

viodragon2 commented on packages/ui/VirtualList/VirtualListBase.js:
```diff
@@ -797,14 +771,44 @@ const VirtualListBaseFactory = (type) => {
+				{props} = this,
+				hasMetricsChanged = (
+					direction !== props.direction ||
+					((itemSize instanceof Object) ? (itemSize.minWidth !== props.itemSize.minWidth || itemSize.minHeight !== props.itemSize.minHeight) : itemSize !== props.itemSize) ||
```
nit: we can use ramda's `equals` function to simplify.

`!equals(prevProps.itemSize, this.props.itemSize)`

ybsung replied:
Oh, it looks much better. Thanks.

ybsung replied:
<On Feb 18th>
- I updated the PR.

viodragon2 commented:
I tried a few things here and there with minimal changes in [this branch](https://github.com/enactjs/enact/tree/feature/PLAT-72723-stephen). Here's a diff you can look at https://github.com/enactjs/enact/compare/feature/PLAT-72723-simple...feature/PLAT-72723-stephen

I'm still struggling how we can handle `prevFirstIndex` and `childProps` before `componentDidUpdate` but hopefully you get a generic idea on how we can move a lot of things in `componentDidUpdate`

ybsung commented on packages/ui/VirtualList/VirtualListBase.js:
```diff
@@ -797,14 +771,44 @@ const VirtualListBaseFactory = (type) => {
+			if (hasMetricsChanged || this.props.clientSize && this.primary === null && this.secondary === null) {
+				this.calculateMetrics(props);
+				this.updateStatesAndBounds(props, true);
+				this.setContainerSize();
```
When `componentDidMount` is called in `VirtualList`, there is no scrollbar even if the content is long enough to scroll it. So the VirtualList re-renders. Then it seems that we need to update the container size again based on the area except the scrollbar.

In addition, when metrics changes or data size changes, we need to update container size again because the visibility of the scrollbar could changes and the scroll area could changes.

ybsung commented:
https://github.com/enactjs/enact/pull/2108#discussion_r256657816
> Is `this.cc` required to be reset when `rtl` prop changes? Doesn't index and key stay the same in RTL layout? Or am I missing something?

As I mentioned in https://github.com/enactjs/enact/pull/2108#discussion_r257102041, let's remove this code to reset `this.cc`.

ybsung commented:
Hi @viodragon2 

I updated the code and the description in the PR. Could you review the PR again?

As you recommended, I moved the code in the deprecated lifecycle methods not to `render` but to `constructor`, `getDerivedStateFromProps`, and `componentDidUpdate`.

But I'm not sure whether it would be better to move the code in `componentDidUpdate` to `getDerivedStateFromProps`. Could you check my comment https://github.com/enactjs/enact/pull/2108#discussion_r257106546 above?
Please let me know your thought.

Thank you so much.

ybsung commented on packages/ui/VirtualList/VirtualListBase.js:
```diff
@@ -222,66 +223,82 @@ const VirtualListBaseFactory = (type) => {
-			}
+		static getDerivedStateFromProps (props, state) {
+			const
+				isNeedToInvalidate = (
```
`shouldInvalidate` would be great. I'll update it.

ybsung commented on packages/ui/VirtualList/VirtualListBase.js:
```diff
@@ -222,66 +223,82 @@ const VirtualListBaseFactory = (type) => {
+			const
+				isNeedToInvalidate = (
+					state.prevFirstIndex === state.firstIndex ||
+					(state.prevChildProps || props.childProps) && state.prevChildProps !== props.childProps
```
I could not remember why I did. The condition that you mentioned should be removed.

ybsung commented on packages/ui/VirtualList/VirtualListBase.js:
```diff
@@ -448,13 +459,16 @@ const VirtualListBaseFactory = (type) => {
 			const
 				{overhang} = props,
-				{firstIndex} = this.state,
+				{firstIndex} = this.state || {firstIndex: 0},
```
`calculateFirstIndex` is called in `constructor` before defining this.state. So there is an error for {firstIndex} = this.state. That is the reason why adding `|| {firstIndex: 0}.

There is another alternative without adding it like the following way.

AS-IS
```
		constructor (props) {
			let nextState = null;

			super(props);

			if (props.clientSize) {
				this.calculateMetrics(props);
				nextState = this.getStatesAndUpdateBounds(this.props);
			}

			this.state = {
				firstIndex: 0,
				numOfItems: 0,
				prevChildProps: null,
				prevFirstIndex: 0,
				updateFrom: 0,
				updateTo: 0,
				...nextState
			};
		}
```

TO-BE
```
		constructor (props) {
			let nextState = null;

			super(props);

			this.state = {
				firstIndex: 0,
				numOfItems: 0,
				prevChildProps: null,
				prevFirstIndex: 0,
				updateFrom: 0,
				updateTo: 0
			};

			if (props.clientSize) {
				this.calculateMetrics(props);
				nextState = this.getStatesAndUpdateBounds(this.props);
			}

			this.state = {
				...this.state,
				...nextState
			};
		}
```
